### ISSUE TITLE
trial reporter doesn't write to streams properly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10995]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1430|
|Type|defect|
|Created|2006-01-19 23:51:22Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1430 1430
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1137714682000000 1137714682000000
changetime__1163895673000000 1163895673000000
version__ 
owner__ 
cc__exarkun cc__jml cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Traceback (most recent call last):
  File "/home/buildbot/run/buildbot/full2.3/Twisted/bin/trial", line 24, in ?
    run()
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/scripts/trial.py",
line 364, in run
    test_result = trialRunner.run(suite)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 465, in run
    suite.run(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 174, in run
    TestSuite.run(self, result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 119, in run
    test(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 111, in __call__
    return self.run(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 119, in run
    test(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 111, in __call__
    return self.run(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 119, in run
    test(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 111, in __call__
    return self.run(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 119, in run
    test(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 111, in __call__
    return self.run(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/runner.py",
line 119, in run
    test(result)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/unittest.py",
line 319, in __call__
    return self.run(*args, **kwargs)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/unittest.py",
line 435, in run
    result.startTest(self)
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/reporter.py",
line 268, in startTest
    self.write('%s ... ', tm.id())
  File "/home/buildbot/run/buildbot/full2.3/Twisted/twisted/trial/reporter.py",
line 135, in write
    self.stream.flush()
IOError: [Errno 4] Interrupted system call
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
If trial is going to write things, it should use t.p.u.untilConcludes.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
This is #733. I didn't realize all the implications before, however. It 
doesn't only affect popen: printing to stdout, when stdout is a terminal rather 
than a file won't work reliably.
Let's kill this thing once and for all by introducing the sigaction call mentioned 
in that bug. Asking people not to print to stdout, or to do strange things when 
doing so isn't a very nice solution.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

```
#!html
<pre>
I can change Trial to use untilConcludes easily. Before that, I'd really a unit
test that consistently fails. How would I go about writing one?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Here's a silly basis for a test:

class AwesomeFile(object):
    lastWrite = None
    def __init__(self, fObj):
        self.fObj = fObj

    def write(self, s):
        if s is self.lastWrite:
            self.lastWrite = s
            return self.fObj.write(s)
        self.lastWrite = None
        raise IOError(errno.EINTR, "Interrupted mumble")

Use this as the stream to log to.

I am inclined to agree with half of James' last comment.  I think we should wrap
sigaction properly and install a signal handler which will not interfere with so
many things.  However, I also don't think trial should be doing blocking writes
directly to standard out.  Perhaps instead of wrapping this one write call in an
untilConcludes, this reporter should be based off of twisted.internet.stdio,
instead.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

I'm disinclined to base the reporter off t.i.stdio.

Regarding sigaction.  I lost all my WRS books when I quit the company that owned them, so I don't really get the signal stuff.  Is this something Trial ought to do?  If so, I will try later to do the research I need to get competent.

If not, then the action for this ticket is to wrap writes in untilConcludes, right?

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

IMO this ticket should be closed as wontfix, and the signal fix applied to core.

I'll try to make that work after I finish the components removal stuff.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Why are you disinclined to use t.i.stdio?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Mostly because I don't understand it, and because it might require me to return Deferreds from all the reporter methods. I might be more inclined if I knew why it was being recommended.

Also, are you saying that this issue will not resolved even if the signal fix is applied?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Stepping back a bit, why do you want this? how is this problem triggered?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

This is still a problem.  The Python 2.3 buildslave fails with this traceback every couple of weeks.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml commented|
|-|-|

Thanks. Upgrading severity.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_15"></a>|@jyknight commented|
|-|-|

Fix in branch for [#10513](https://github.com/twisted/twisted/issues/10513) ought to fix this too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

[#10513](https://github.com/twisted/twisted/issues/10513) removes signal.signal(signal.SIGCHLD, ...) from Twisted, which will fix this issue for SIGCHLD.  What about other signals?  For example, a SIGUSR1 handler is installed by twistd.  Can this potentially cause an EINTR?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_17"></a>|@jml set owner to @exarkun|
|-|-|

Ready for review in reporter-streams-1430

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @jml|
|-|-|

MinimalReporter's printSummary method calls write on self.stream directly, which will bypass the changes made to Reporter.write.  Ditto for _AnsiColorizer and _NullColorizer (and _Win32Colorizer, but I don't think that one matters), since it is passed the underlying stream, rather than a wrapper.

TrialReporter.runUntilFailure also writes directly to its stream object.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_19"></a>|@jml set owner to @exarkun|
|-|-|

OK, I tried again with a different approach. 

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @jml|
|-|-|

Okay, I did finally succeed in getting .flush() to raise an IOError.  It would probably make sense to move the untilConcludes(flush) into SafeStream and have the reporter just call flush, as it just calls write.  I think I'd also feel better if the __getattr__ on SafeStream were dropped: it makes it too easy to accidentally do something that may not work.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_21"></a>|@jml set status to `closed`|
|-|-|

From r18719

Make Trial reporters safely write to streams.
 * Author: jml
 * Reviewer: exarkun
 * Fixes: [#10995](https://github.com/twisted/twisted/issues/10995)

Wraps the calls to `write` and `flush` in `untilConcludes`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
